### PR TITLE
Persistence with skopt.dump() and skopt.load()

### PR DESCRIPTION
**Pull request by mp4096**
See #190 

To do:
- [x] Write unit tests

~~Be more specific when catching the pickling exception, search for `func` in the error message~~

### CODE REVIEW

**Comment by betatim** [comment]
Rendered notebook: https://173-54340642-gh.circle-artifacts.com/0/tmp/circle-artifacts.jN1zNtS/doc/skopt/notebooks/store-and-load-results.html

**Comment by betatim** [comment]
I am in favour of providing these two methods for users (as opposed to #246). Mainly to save on typing.

* Should we use some magic to try and import `dill` and if it works use that instead? I am -1
* Instead of removing `func` and retrying the pickling I would add an argument to, maybe `dump(res, without_objective=True)`, where the user can explicitly say that they want to store things without the objective

Thoughts @glouppe and @MechCoder?

**Comment by mp4096** [comment]
I'm against trying to import `dill` and using `pickle` as a fallback due to the compatibility. The `load()` function cannot know which one was used; in worst case, one has to try both.

**Comment by MechCoder** [comment]
Thanks for handling this. We should also have some sort of checkpointing machinery so that the information is dumped after every iteration incase something goes wrong after some iteration.

**Comment by betatim** [comment]
Coming back to this, I think we want this approach and not the other one in #246. Things that need doing are tracked in the first post of this PR.

We decided against `dill` (and guessing). I would also prefer that we simplify things by asking the user to explicitly decide if they [want to store the objective or not](https://github.com/scikit-optimize/scikit-optimize/pull/245#issuecomment-252556914). This will make our code cleaner and no one will be surprised if there is no objective in the pickle.

I'd put checkpointing in a new PR once this is merged. Does that work for you @MechCoder?

Would be good to merge this before 0.2 is released.

**Comment by mp4096** [comment]
Ok,, I'll add an explicit argument for storing the objective function!

File: skopt/utils.py [hunk 1]
```diff
@@ -44,3 +52,58 @@ def create_result(Xi, yi, space=None, rng=None, specs=None, models=None):
...
+    * `res` [`OptimizeResult`, scipy object]:
+        Optimization result object to be stored.
+
+    * `filename` [string]:
+        The path of the file in which it is to be stored.
+    """
+
+    try:
+        with open(filename, 'wb') as f:
+            pickle.dump(res, f)
```
Comment (MechCoder) [inline skopt/utils.py]: We should use joblib.dump no? (from `sklearn.externals.joblib`)

File: examples/store-and-load-results.ipynb [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+   "metadata": {},
+   "source": [
+    "`skopt.dump()` and `skopt.load()` use the `pickle` module internally. Notice that if the objective function is not serializable, `skopt.dump()` will raise a warning and store a copy of the optimization results object without the objective function. \n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Custom alternatives\n",
```
Comment (MechCoder) [inline examples/store-and-load-results.ipynb]: Is this section necessary? Not sure we should be explaining about pickle or joblib here (especially since we are not doing anything more other than a thin wrapper over these)

**Comment by MechCoder** [comment]
What do you guys think about adding an argument to the minimize functionalities that stores the optimisation results on to disk directly after the optimisation is over?

**Comment by betatim** [comment]
A bit weary of adding yet more arguments to the `*_minimize()` function. At some point it will be able to configure the coffee machine to brew coffee while we are optimising :-/ I'd rather support having a ready made callback that stores the result every `n` iterations.

File: skopt/utils.py (hunk 1 above)
Comment (betatim) [inline skopt/utils.py]: Somewhere we discussed whether or not to always use `joblib`. I think it was about "is it necessary to use `joblib` or would `pickle` suffice?" Can't find that thread right now. I'd be happy simply using `joblib` always.

File: examples/store-and-load-results.ipynb (hunk 2 above)
Comment (mp4096) [inline examples/store-and-load-results.ipynb]: Ok, I'll remove it.

File: skopt/utils.py (hunk 1 above)
Comment (mp4096) [inline skopt/utils.py]: It was discussed in #190, somewhere around [this comment](https://github.com/scikit-optimize/scikit-optimize/issues/190#issuecomment-249296787)

I'm ok with `joblib`.

**Comment by mp4096** [comment]
Liked your comment about the coffee brewing. An email notification after each iteration would be nice too. :grin:

I think a checkpoint callback is an elegant solution.

**Comment by betatim** [comment]
> Liked your comment about the coffee brewing. An email notification after each iteration would be nice too. :grin:

-> in a new PR please :)

**Comment by mp4096** [comment]
* Now using `joblib`. Additional keyword arguments are simply propagated further to `joblib.dump()` / `joblib.load()` 
* I think `store_objective` is better than `without_objective` for it is positive/affirmative. Maybe `dump_objective` would be even better?

**Comment by mp4096** [comment]
I wasn't sure whether the unit test for `dump()` and `load()` should go into `test_common.py` or in its own script. Is it ok this way?

File: skopt/tests/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+from skopt import dump
+
+
+def obj_fun(x):
+    return np.sin(5 * x[0]) * (1 - np.tanh(x[0] ** 2))
+
+
+def check_optimization_results_equality(res_1, res_2):
+    # Check if the results objects have the same keys
+    assert_equal(res_1.keys(), res_2.keys())
```
Comment (MechCoder) [inline skopt/tests/test_utils.py]: Should sort before comparing?

File: skopt/tests/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+from sklearn.utils.testing import assert_array_equal
+from sklearn.utils.testing import assert_array_less
+from sklearn.utils.testing import assert_equal
+from sklearn.utils.testing import assert_not_equal
+from sklearn.utils.testing import assert_less_equal
+from sklearn.utils.testing import assert_greater
+from sklearn.utils.testing import assert_true
+
+from skopt import gp_minimize
+
```
Comment (MechCoder) [inline skopt/tests/test_utils.py]: extra space

File: skopt/tests/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        dump(res, f)
+        f.seek(0)
+        res_loaded = load(f)
+    check_optimization_results_equality(res, res_loaded)
+    assert_true('func' in res_loaded.specs['args'])
+
+    # Test dumping without objective function
+    with tempfile.TemporaryFile() as f:
+        dump(res, f, store_objective=False)
+        f.seek(0)
```
Comment (MechCoder) [inline skopt/tests/test_utils.py]: Why is this line necessary?

File: skopt/tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+from sklearn.utils.testing import assert_greater
+from sklearn.utils.testing import assert_true
+
+from skopt import gp_minimize
+
+from skopt import load
+from skopt import dump
+
+
+def obj_fun(x):
```
Comment (MechCoder) [inline skopt/tests/test_utils.py]: You can import from `skopt.benchmarks` but meh...

File: skopt/utils.py [hunk 7]
```diff
@@ -44,3 +51,70 @@ def create_result(Xi, yi, space=None, rng=None, specs=None, models=None):
...
+        method corresponding to one of the supported filename extensions ('.z',
+        '.gz', '.bz2', '.xz' or '.lzma') will be used automatically.
+
+    * `store_objective` [boolean, default=True]:
+        Whether the objective function should be stored. Set `store_objective`
+        to `False` if your objective function (`.specs['args']['func']`) is
+        unserializable (i.e. if an exception is raised when trying to serialize
+        the optimization result). Notice that in this case a deep copy of the
+        optimization result is created. This can potentially lead to performance
+        problems.
```
Comment (MechCoder) [inline skopt/utils.py]: You can add the note that it is advisable to delete the objective function before dumping to avoid such issues

**Comment by MechCoder** [comment]
Looks fine pending minor comments. Thanks @mp4096! @betatim please merge if happy.

File: skopt/tests/test_utils.py (hunk 5 above)
Comment (mp4096) [inline skopt/tests/test_utils.py]: * A `tempfile.TemporaryFile` object is deleted right after it has been closed, in this case when we leave the `with` context.
* `joblib` writes into the file and leaves the seeker pointing to EOF. So if we `read` or `load` it, the file will appear to be empty.
* So we have to 'rewind' the file object to the beginning.

File: skopt/tests/test_utils.py (hunk 6 above)
Comment (mp4096) [inline skopt/tests/test_utils.py]: Good point, thanks!

File: skopt/tests/test_utils.py (hunk 5 above)
Comment (MechCoder) [inline skopt/tests/test_utils.py]: Are you sure about the second point? Atleast tests pass locally after I remove the `seek` line.

File: skopt/tests/test_utils.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import tempfile
+
+from sklearn.utils.testing import assert_array_almost_equal
+from sklearn.utils.testing import assert_array_equal
+from sklearn.utils.testing import assert_array_less
+from sklearn.utils.testing import assert_equal
+from sklearn.utils.testing import assert_not_equal
+from sklearn.utils.testing import assert_less_equal
+from sklearn.utils.testing import assert_greater
+from sklearn.utils.testing import assert_true
```
Comment (betatim) [inline skopt/tests/test_utils.py]: could you remove the imports that aren't used please? Needed: assert_true, assert_equal and assert_array_equal

File: skopt/tests/test_utils.py (hunk 5 above)
Comment (mp4096) [inline skopt/tests/test_utils.py]: You're absolutely right, sorry! I just checked that `dump` leaves the seeker at the EOF and haven't thought that `load` itself might do `seek(0)`, which it does.

File: skopt/tests/test_utils.py (hunk 8 above)
Comment (mp4096) [inline skopt/tests/test_utils.py]: Thanks, missed this one.

**Comment by mp4096** [comment]
Are you ok with the wording in the example notebook?
https://github.com/mp4096/scikit-optimize/blob/dev-save/examples/store-and-load-results.ipynb

**Comment by betatim** [comment]
Merged! 🎉 🎈

**Comment by mp4096** [comment]
Cool, thanks! :cake:
